cmake/introspection.cmake

```diff
@@ -0,0 +1,281 @@
+  check_cxx_symbol_exists(le16toh ${ENDIAN_HEADER} HAVE_DECL_LE16TOH)
+  check_cxx_symbol_exists(le32toh ${ENDIAN_HEADER} HAVE_DECL_LE32TOH)
+  check_cxx_symbol_exists(le64toh ${ENDIAN_HEADER} HAVE_DECL_LE64TOH)
+endif()
+
+check_include_file_cxx(unistd.h HAVE_UNISTD_H)
+if(HAVE_UNISTD_H)
+  check_cxx_symbol_exists(fdatasync "unistd.h" HAVE_FDATASYNC)
+  check_cxx_symbol_exists(fork "unistd.h" HAVE_DECL_FORK)
+  check_cxx_symbol_exists(pipe2 "unistd.h" HAVE_DECL_PIPE2)
+  check_cxx_symbol_exists(setsid "unistd.h" HAVE_DECL_SETSID)
+endif()
+
+check_include_file_cxx(sys/types.h HAVE_SYS_TYPES_H)
+check_include_file_cxx(ifaddrs.h HAVE_IFADDRS_H)
+if(HAVE_SYS_TYPES_H AND HAVE_IFADDRS_H)
+  check_cxx_symbol_exists(freeifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_FREEIFADDRS)
+  check_cxx_symbol_exists(getifaddrs "sys/types.h;ifaddrs.h" HAVE_DECL_GETIFADDRS)
+  # Illumos/SmartOS requires linking with -lsocket if
+  # using getifaddrs & freeifaddrs.
+  # See: https://github.com/bitcoin/bitcoin/pull/21486
+  if(HAVE_DECL_GETIFADDRS AND HAVE_DECL_FREEIFADDRS)
+    set(check_socket_source "
+      #include <sys/types.h>
+      #include <ifaddrs.h>
+
+      int main() {
+        struct ifaddrs* ifaddr;
+        getifaddrs(&ifaddr);
+        freeifaddrs(ifaddr);
+      }
+    ")
+    check_cxx_source_links("${check_socket_source}" IFADDR_LINKS_WITHOUT_LIBSOCKET)
+    if(NOT IFADDR_LINKS_WITHOUT_LIBSOCKET)
+      check_cxx_source_links_with_libs(socket "${check_socket_source}" IFADDR_NEEDS_LINK_TO_LIBSOCKET)
+      if(IFADDR_NEEDS_LINK_TO_LIBSOCKET)
+        link_libraries(socket)
+      else()
+        message(FATAL_ERROR "Cannot figure out how to use getifaddrs/freeifaddrs.")
+      endif()
+    endif()
+  endif()
+endif()
+
+# Check for gmtime_r(), fallback to gmtime_s() if that is unavailable.
+# Fail if neither are available.
+check_cxx_source_compiles("
+  #include <ctime>
+
+  int main()
+  {
+    gmtime_r((const time_t*)nullptr, (struct tm*)nullptr);
+  }
+  " HAVE_GMTIME_R
+)
+if(NOT HAVE_GMTIME_R)
+  check_cxx_source_compiles("
+    #include <ctime>
+
+    int main()
+    {
+      gmtime_s((struct tm*)nullptr, (const time_t*)nullptr);
+    }
+    " HAVE_GMTIME_S
+  )
+  if(NOT HAVE_GMTIME_S)
+    message(FATAL_ERROR "Both gmtime_r and gmtime_s are unavailable.")
+  endif()
+endif()
+
+check_cxx_source_compiles("
+  #include <cstdlib>
+
+  int main()
+  {
+    return std::system(nullptr);
+  }
+  " HAVE_STD_SYSTEM
+)
+check_cxx_source_compiles("
+  #include <stdlib.h>
+
+  int main()
+  {
+    return ::_wsystem(NULL);
+  }
+  " HAVE__WSYSTEM
+)
+if(HAVE_STD_SYSTEM OR HAVE__WSYSTEM)
+  set(HAVE_SYSTEM 1)
+endif()
+
+check_include_file_cxx(string.h HAVE_STRING_H)
+if(HAVE_STRING_H)
+  check_cxx_source_compiles("
+    #include <string.h>
+
+    int main()
+    {
+      char buf[100];
+      char x = *strerror_r(0, buf, sizeof buf);
+      char* p = strerror_r(0, buf, sizeof buf);
+      return !p || x;
+    }
+    " STRERROR_R_CHAR_P
+  )
```

Comment:
I took it from https://github.com/autotools-mirror/autoconf/blob/9d2637467df1d4811449dd14b2bebc142d9092b6/lib/autoconf/functions.m4#L1707-L1738.

Nevertheless, I'm going to consider and test your suggestion.

Reply:
Hmm,
```c++
static_assert(std::is_same<decltype(strerror_r), char*(int, char*, size_t)>::value);
```
looks fragile as `noexcept` could be a part of the function type.

Reply:
Then it could be
```cpp
static_assert(std::is_same<decltype(strerror_r), char*(int, char*, size_t)>::value ||
              std::is_same<decltype(strerror_r), char*(int, char*, size_t) noexcept>::value);
```

Reply:
> 1\. Why does the `return` value matter since `check_cxx_source_compiles()` is used?

To ["avoid unused variable warning"](https://github.com/autotools-mirror/autoconf/commit/3ddd7fba50272221ad09d9ae044b9891965fc122).

> 2\. Why call `strerror_r()` two times?

I guess that the first call is mostly for C language. Indeed, it is redundant in Bitcoin Core.

Reply:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1441845083).

Reply:
> Also, I think both calls are equally C-ish or C++-ish.

I don't think that C compiler will fail if the return type of `func()` is a pointer to some other type.